### PR TITLE
Implemented dependency injection for the data product endpoints

### PR DESCRIPTION
**Pull request by vladimir-gladkii**
| :rocket: | This description was created by [Ellipsis](https://www.ellipsis.dev) for commit f0361bb2d2cfd4c7dea745131ebb944cdece0ab9  | 
|--------|--------|

### Summary:
Implemented dependency injection in `DataProductRoutes` using `IDataProductUsecases` and `DataProductUsecases`, enhancing testability and flexibility.

**Key points**:
- Implemented dependency injection in `DataProductRoutes` for data product endpoints.
- Introduced `IDataProductUsecases` interface and `DataProductUsecases` class for use case injection.
- Modified `DataProductRoutes` to accept `IDataProductUsecases` in constructor.
- Replaced direct `usecases` calls with `self._usecases` in methods.
- Updated `routes` initialization to use `DataProductUsecases`.
- Refactored tests to use mock use cases.
- Added `create_client` and `create_usecases` helper functions for test setup.
- Ensured tests cover scenarios for successful data retrieval and exceptions.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: server/app/rest_api/routes/data_products.py [hunk 1]
```diff
@@ -27,7 +38,7 @@ async def get_data_products(
     ) -> PaginatedResult:
         """Returns a list of data products with the ability to paginate"""
         query = queries.GetPaginated(page=page, size=pageSize)
-        data_products = await usecases.get_data_products_list(query)
+        data_products = await self._usecases["list"](query)
```
Comment (alishalbaf) [inline server/app/rest_api/routes/data_products.py]: you can define interface 
```
class ProductServiceInterface:
@abstractmethod
    def list(self,query):
```

and have an implementation of that

```
class ProductService:
    def list(self,query):

```
and in your routable have just 
```
class DataProductRoutes:
    def __init__(self, product_service: DataProductServiceInterface):
        self.DataProduct_service = product_service
```

in this way you are fixed the methods that are available in your service (usecase) .

Comment (vladimir-gladkii) [reply server/app/rest_api/routes/data_products.py]: You are right. To define the ProductService is a good idea.
Do you think it is a good idea to define an interface in this case? We will have only one implementation.

Comment (vladimir-gladkii) [reply server/app/rest_api/routes/data_products.py]: I prefer to use interfaces for classes which can have many implementations to reduce the complexity. @alishalbaf @RafyMehany what do you think should we define an interface for class that has only one implementation?

Comment (alishalbaf) [reply server/app/rest_api/routes/data_products.py]: in this case you have defined a dictionary of string to callable , in this way you prevent static analyser to check the code for possible wrong accessing the methods : self._usecases["list"]. if you define interface one of its advantages is to allow static analysers check the code and reduce the errors

Comment (vladimir-gladkii) [reply server/app/rest_api/routes/data_products.py]: I will have the same result (for static analyser) if I define class without the interface.
I understand the advantages of interfaces. I just want to align rules for code.
I think we should discuss it.

Comment (vladimir-gladkii) [reply server/app/rest_api/routes/data_products.py]: Fixed

**Review by alishalbaf — APPROVED** [review]

**Pull request merged**